segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/IndexRequestProcessor.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.segmentstore.server.host.handler;
+
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.SortUtils;
+import io.pravega.segmentstore.contracts.ReadResult;
+import io.pravega.segmentstore.contracts.ReadResultEntry;
+import io.pravega.segmentstore.contracts.SegmentProperties;
+import io.pravega.segmentstore.contracts.StreamSegmentStore;
+import io.pravega.shared.NameUtils;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Map.Entry;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A Process for all index segment related operations.
+ */
+@Slf4j
+public final class IndexRequestProcessor {
+    private static final Duration TIMEOUT = Duration.ofMinutes(1);
+
+    static final class SegmentTruncatedException extends RuntimeException {
+        private static final long serialVersionUID = 1L;
+
+        public SegmentTruncatedException(String message) {
+            super(message);
+        }
+    }
+
+    /**
+     * Locate the requested offset in segment.
+     *
+     * @param store  The StreamSegmentStore to attach to (and issue requests to).
+     * @param segment Segment name.
+     * @param targetOffset The requested offset's corresponding position to search in the index segment entry.
+     * @param greater boolean to determine if the next higher or the lower value to be returned in case the requested offset is not present.
+     *
+     * @return the corresponding offset position from the index segment entry.
+     * @throws SegmentTruncatedException If the segment is truncated.
+     */
+    public static long locateOffsetForSegment(StreamSegmentStore store, String segment, long targetOffset,
+                                              boolean greater) throws SegmentTruncatedException {
+        String indexSegmentName = NameUtils.getIndexSegmentName(segment);
+
+        // Fetch start and end idx.
+        SegmentProperties properties = store.getStreamSegmentInfo(indexSegmentName, TIMEOUT).join();
+        long startIdx = properties.getStartOffset() / NameUtils.INDEX_APPEND_EVENT_SIZE;
+        long endIdx = properties.getLength() / NameUtils.INDEX_APPEND_EVENT_SIZE - 1;
+        // If startIdx and endIdx are same, then pass length of segment as a result.
+
+        if (startIdx > endIdx) {
+            SegmentProperties segmentProperties = store.getStreamSegmentInfo(segment, TIMEOUT).join();
```

Comment:
Humm, I suppose it is public, so in theory it could be called from a core thread. 
It is not now. But I can add that it must not be called from a core thread in the JavaDoc. 
Longer term we probably want to make this async, this of course involves making the whole of the search async as it need to fetch items from the index.